sdk/java/pulumi/src/main/java/com/pulumi/core/Urn.java

```diff
@@ -1,75 +1,312 @@
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(stack, project, qualifiedType, name);
+    }
 
-    private Urn() {
-        throw new UnsupportedOperationException("static class");
+    @Override
+    public String toString() {
+        return new StringJoiner(", ", Urn.class.getSimpleName() + "[", "]")
+                .add("stack='" + stack + "'")
+                .add("project='" + project + "'")
+                .add("qualifiedType=" + qualifiedType)
+                .add("name='" + name + "'")
+                .toString();
+    }
+
+    @InternalUse
+    @SuppressWarnings("ConstantConditions") // IntelliJ can't understand the custom validators
+    public static Urn parse(String urn) {
+        require(u -> isNonEmptyOrNull(u), urn,
+                () -> format("expected urn to be not empty and not null, got: '%s'", urn)
+        );
+        require(u -> u.startsWith(Prefix), urn,
+                () -> format("expected urn to start with '%s', got: '%s'", Prefix, urn)
+        );
+
+        var urnParts = urn.split(PartsSeparator);
+        require(up -> up.length == 4, urnParts,
+                () -> format("expected urn to have 4 parts, separated by '%s', got '%s' in '%s'",
+                        PartsSeparator, urnParts.length, urn
+                )
+        );
+        var stack = urnParts[0].replaceFirst(Prefix, "");
+        require(s -> isNonEmptyOrNull(s), stack,
+                () -> format("expected urn stack part to be not empty, got: '%s'", stack)
+        );
+        require(s -> !s.contains(PartsSeparator), stack,
+                () -> format("expected urn stack part to not contain '%s', got: '%s'", ParentSeparator, stack)
+        );
+        var project = urnParts[1];
+        require(p -> isNonEmptyOrNull(p), project,
+                () -> format("expected urn project part to be not empty, got: '%s'", project)
+        );
+        require(p -> !p.contains(PartsSeparator), project,
+                () -> format("expected urn project part to not contain '%s', got: '%s'", ParentSeparator, project)
+        );
+        var qualifiedType = urnParts[2];
+        require(q -> isNonEmptyOrNull(q), qualifiedType,
+                () -> format("expected urn qualifiedType part to be not empty, got: '%s'", qualifiedType)
+        );
+        var name = urnParts[3];
+        require(n -> isNonEmptyOrNull(n), name,
+                () -> format("expected urn name part to be not empty, got: '%s'", name)
+        );
+
+        return new Urn(stack, project, QualifiedTypeName.parse(qualifiedType), name);
+    }
+
+    public String asString() {
+        return Prefix + this.stack + PartsSeparator
+                + this.project + PartsSeparator
+                + this.qualifiedType.asString() + PartsSeparator
+                + this.name;
+    }
+
+    @ParametersAreNonnullByDefault
+    @SuppressWarnings("OptionalUsedAsFieldOrParameterType")
+    public static final class QualifiedTypeName {
```

Comment:
Is it useful to have this and Type public?

Reply:
they are used in various packages of SDK, we could annotate all 3 classes with `@InternalUse`, I don't think we're exposing Urn to the user anyway, the providers return Strings

Reply:
Sounds good then!

Reply:
done, also moved to internal package